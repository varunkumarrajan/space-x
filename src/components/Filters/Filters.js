// React
import React from 'react';
// Prop types
import PropTypes from 'prop-types'; // ES6
// Styled component
import styled from 'styled-components';
// Components
import Button from 'components/Button/Button';
// Device
import device from 'config/BreakPoints';
// Constant
import constant from 'constant/Constant';

// Destructuring
const { LATENCY } = constant;

// Styled container
const Container = styled.div`
  flex: 0 0 260px;
  @media screen and ${device.mobile} {
    flex: 0 0 100%;
  }
`;

// Styled filter wrapper
const FiltersWrapper = styled.section`
  background-color: #ffffff;
  margin: 10px;
  padding: 10px;
  height: fit-content;
`;

// Styled main heading
const MainHeading = styled.h4`
  font-weight: bolder;
  font-size: 20px;
`;

// Styled subheading
const SubHeading = styled.div`
  text-align: center;
  border-bottom: 1px solid #ddd;
  margin: 0 20px;
`;

// Styled divider
const Divider = styled.div`
  margin: 5px 0;
  @media screen and ${device.mobile} {
    text-align: center;
  }
`;

/**
 * @description Filters Element
 * @param {object} { content, handleFilter}
 */
const Filters = ({ content, handleFilter }) => {
  // Local state
  const [event, setEvent] = React.useState({click: 0, filterKey: '', filterType: ''});

  // Update check click & double click
  React.useEffect(() => {
    let timer;
    if (event.click === 1) {
      timer = setTimeout(
        () => {
          handleFilters();
          setEvent({...event, click: 0});
        }, LATENCY);
    } else if (event.click === 2) {
      handleFilters(true);
      setEvent({...event, click: 0});
    }
    return () => clearTimeout(timer);
  }, [event.click]);

  /**
   * @description handle click
   * @param {string} filterKey
   * @param {string} filterType
   */
  const handleClick = (filterKey, filterType) => setEvent({filterKey, filterType, click: 1})

  /**
   * @description handle double click
   * @param {string} filterKey
   * @param {string} filterType
   */
  const handleDoubleClick = (filterKey, filterType) => setEvent({filterKey, filterType, click: 2});

  /**
   * @description
   * @param {boolean} [isCheck=false]
   */
  const handleFilters = (isCheck = false) => {
    // get key
    let updatedFilter = event.filterKey;
    // check update filter or not
    if (isCheck)
      // Loop over the filters
      content.filters.forEach((filter) => {
        // check filter type (ex: year)
        if (filter.key === event.filterType) {
          // update filter with blank
          updatedFilter = '';
        }
      });
    // call the callback
    handleFilter(
      content.filters.map((f) =>
        f.key === event.filterType ? { ...f, value: updatedFilter } : f
      )
    );
  };

  // return filter component (html)
  return (
    <Container>
      <FiltersWrapper>
        <MainHeading>{content.title}</MainHeading>
        {content.filterTypes.map((filterType, ind) => (
          <React.Fragment key={ind}>
            <SubHeading>{filterType.title}</SubHeading>
            <Divider>
              {content.filterDetails[filterType.key].map((filter, i) => (
                <Button
                  key={i}
                  handleButtonClick={handleClick}
                  handleButtonDoubleClick={handleDoubleClick}
                  button={{
                    ...filter,
                    type: filterType.key,
                    defaultValue: content.filters.filter(
                      (f) => f.key === filterType.key
                    )[0].value,
                  }}
                />
              ))}
            </Divider>
          </React.Fragment>
        ))}
      </FiltersWrapper>
    </Container>
  );
};

// Default export
export default Filters;

// Default props
Filters.defaultProps = {
  content: {},
  handleFilter: () => {},
};

// Prop types
Filters.propTypes = {
  content: PropTypes.any.isRequired,
  handleFilter: PropTypes.func.isRequired,
};
