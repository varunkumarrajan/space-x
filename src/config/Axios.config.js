// Axios
import axios from 'axios';
// Constant
import constant from 'constant/Constant';

// Constant destructuring
const { baseURL, timeout } = constant;

// Axios config
const axiosConfig = axios.create({ baseURL, timeout });

// Default export
export default axiosConfig;
