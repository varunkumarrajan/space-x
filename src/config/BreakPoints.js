// BreakPoints
const breakPoints = {
  mobile: '700px',
  tabletN: '701px',
  tabletX: '1024px',
  desktopN: '1025px',
  desktopX: '1440px',
  desktopL: '1441px',
};

// Device
const device = {
  mobile: `(max-width: ${breakPoints.mobile})`,
  tablet: `(min-width: ${breakPoints.tabletN}) and (max-width: ${breakPoints.tabletX})`,
  desktop: `(min-width: ${breakPoints.desktopN}) and (max-width: ${breakPoints.desktopX})`,
  desktopL: `(min-width: ${breakPoints.desktopL})`,
};

// Default export
export default device;
