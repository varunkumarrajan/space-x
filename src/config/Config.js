// Labels
import labels from './Labels';

// Destructuring
const { filters } = labels;

// Config
const config = {
  launchSuccess: (launch) => (launch ? `&launch_success=${launch}` : ``),
  landSuccess: (land) => (land ? `&land_success=${land}` : ``),
  launchYear: (year) => (year ? `&launch_year=${year}` : ``),
  apiURL: {
    launchQuery: (launch, land, year) =>
      `launches?limit=100${config.launchSuccess(launch)}${config.landSuccess(
        land
      )}${config.launchYear(year)}`,
  },
  filterTypes: [
    { key: 'years', title: filters.year, value: '', multi: true },
    { key: 'launch', title: filters.launch, value: '', multi: false },
    { key: 'land', title: filters.land, value: '', multi: false },
  ],
  filterDetails: {
    years: [
      { key: '2006', value: '2006' },
      { key: '2007', value: '2007' },
      { key: '2008', value: '2008' },
      { key: '2009', value: '2009' },
      { key: '2010', value: '2010' },
      { key: '2011', value: '2011' },
      { key: '2012', value: '2012' },
      { key: '2013', value: '2013' },
      { key: '2014', value: '2014' },
      { key: '2015', value: '2015' },
      { key: '2016', value: '2016' },
      { key: '2017', value: '2017' },
      { key: '2018', value: '2018' },
      { key: '2019', value: '2019' },
      { key: '2020', value: '2020' },
    ],
    launch: [
      { key: 'true', value: 'True' },
      { key: 'false', value: 'False' },
    ],
    land: [
      { key: 'true', value: 'True' },
      { key: 'false', value: 'False' },
    ],
  },
};

// Default config
export default config;
