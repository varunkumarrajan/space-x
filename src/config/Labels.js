// Labels
const labels = {
  layout: {
    header: {
      heading: 'SpaceX Launch Programs',
    },
    footer: {
      title: 'Developed By',
      name: 'Varun Kumar Rajan',
    },
  },
  filters: {
    title: 'Filters',
    missionId: 'Mission Ids',
    year: 'Launch Year',
    launch: 'Successful Launch',
    land: 'Successful Landing',
  },
  noContent: 'No Records Found',
};

// Default export
export default labels;
