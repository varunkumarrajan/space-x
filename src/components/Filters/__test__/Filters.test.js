// React
import React from 'react';
// Testing library
import { render, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';

// Labels
import labels from 'config/Labels';
// Config
import config from 'config/Config';

import Filters from './../Filters';

// Destructuring
const {
  filterDetails,
  filterTypes,
} = config;

// Mock data
const filtersMock = {
  title: labels.filters.title,
  filterDetails,
  filterTypes,
  filters: filterTypes,
}

const handleFilter = () => {}

afterEach(cleanup);

describe('Filters', () => {
  it('Should Render Filters With Props', () => {
    const { container } = render(<Filters content={filtersMock} handleFilter={handleFilter} />);
    expect(container.children).not.toHaveLength(0);
  });

  it('Filters Snapshot', () => {
    const filters = renderer.create(<Filters content={filtersMock} handleFilter={handleFilter} />).toJSON();
    expect(filters).toMatchSnapshot();
  });
});
