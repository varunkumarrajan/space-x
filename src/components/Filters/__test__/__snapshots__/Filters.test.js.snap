// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters Filters Snapshot 1`] = `
<div
  className="sc-gsTCUz eDbDeT"
>
  <section
    className="sc-dlfnbm jsrkcH"
  >
    <h4
      className="sc-hKgILt efvMBG"
    >
      Filters
    </h4>
    <div
      className="sc-eCssSg eidbcA"
    >
      Launch Year
    </div>
    <div
      className="sc-jSgupP eWnIQF"
    >
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2006
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2007
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2008
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2009
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2010
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2011
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2012
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2013
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2014
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2015
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2016
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2017
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2018
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2019
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        2020
      </button>
    </div>
    <div
      className="sc-eCssSg eidbcA"
    >
      Successful Launch
    </div>
    <div
      className="sc-jSgupP eWnIQF"
    >
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        True
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        False
      </button>
    </div>
    <div
      className="sc-eCssSg eidbcA"
    >
      Successful Landing
    </div>
    <div
      className="sc-jSgupP eWnIQF"
    >
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        True
      </button>
      <button
        className="sc-bdfBwQ hbzmUa"
        data-testid="buttonId"
        onClick={[Function]}
      >
        False
      </button>
    </div>
  </section>
</div>
`;
